<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [977]:
import pandas as pd
import psycopg2

In [978]:
# Параметры подключения из юнита 1. Работа с базой данных из Python
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

Памятка: Не забудьте удалить ячейку с данными соединения перед фиксацией работы в GitHub.

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [980]:
# Количество вакансий в базе.
query_3_1 = f'''select
    count(*)
    from public.vacancies
'''
display(pd.read_sql_query(query_3_1, connection))

C:\Users\Leon\AppData\Local\Temp\ipykernel_3528\700927336.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  display(pd.read_sql_query(query_3_1, connection))


count
0  49197

__Вывод: в базе данных находятся 49197 вакансий.__

***

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [981]:
# Количество работодателей.
query_3_2 = f'''select
    count(*)
    from public.employers
'''
display(pd.read_sql_query(query_3_2, connection))

C:\Users\Leon\AppData\Local\Temp\ipykernel_3528\1035537684.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  display(pd.read_sql_query(query_3_2, connection))


count
0  23501

__Вывод: количество работодателей составляет 23501.__

***

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [982]:
# Количество регионов
query_3_3 = f'''select
    count(*)
    from public.areas
'''
display(pd.read_sql_query(query_3_3, connection))

C:\Users\Leon\AppData\Local\Temp\ipykernel_3528\3950709734.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  display(pd.read_sql_query(query_3_3, connection))


count
0   1362

__Вывод: в таблице areas число регионов составляет 1362.__

***

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [983]:
# Количество сфер деятельности.
query_3_4 = f'''select
    count(*)
    from public.industries
'''
display(pd.read_sql_query(query_3_4, connection))

C:\Users\Leon\AppData\Local\Temp\ipykernel_3528\3030964408.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  display(pd.read_sql_query(query_3_4, connection))


count
0    294

__Вывод: количество сфер деятельности в базе составляет 294.__

***

#### __Выводы по предварительному анализу данных__
__В базе данных представлены 49197 вакансий. Количество работодателей равно 23501. Количество регионов, в которых представлены вакансии, состовляет 1362 региона и они затрагивают 294 сферы деятельности.__

.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [984]:
# Количество вакансий по регионам.
query_4_1 = f'''select
    a.name, -- Названия регионов.
    count(v.id) cnt, -- Подсчет количества вакансий по регионам.
    v.area_id area -- Номер региона.
from public.vacancies v -- Таблица с вакансиями.
    join public.areas a on v.area_id = a.id -- Присоединение таблицы с регионами к таблице с вакансиями.
group by v.area_id, a.name -- Группировка по регионам и по их названию.
order by cnt desc -- Сортировка по количеству вакансий.
limit 5 -- Ограничение строк вывода в соответствии с заданием.
'''
display(pd.read_sql_query(query_4_1, connection)) # Вывод данных.

C:\Users\Leon\AppData\Local\Temp\ipykernel_3528\330023882.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  display(pd.read_sql_query(query_4_1, connection)) # Вывод данных.


name   cnt  area
0           Москва  5333     1
1  Санкт-Петербург  2851     2
2            Минск  2112  1002
3      Новосибирск  2006     4
4           Алматы  1892   160

__Вывод: в пятерку лидеров по количеству вакансий входят Москва, Санкт-Петербург, Минск, Новосибирск и Алматы.__

***

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [985]:
# Количество вакансий.
query_4_2 = f'''select
    count(*)
from public.vacancies
where (salary_from is not null) or (salary_to is not null) -- Фильтрация по наличию хотя бы одного поля с зарплатой.
'''
display(pd.read_sql_query(query_4_2, connection))

C:\Users\Leon\AppData\Local\Temp\ipykernel_3528\771204417.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  display(pd.read_sql_query(query_4_2, connection))


count
0  24073

__Вывод: количество вакансий, где указано хотя бы одно поле с зарплатой, состовляет 24073. Можно сделать вывод, что у половины вакансий требования не очень жесткие по отношению к зарплате.__

***

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [986]:
# Средние значения нижней и верхней границ зарплаты.
query_4_3 = f'''select
    round(avg(salary_from)) average_minimum, -- Округление среднего значения минимальной заработной платы.
    round(avg(salary_to)) average_maximum
from public.vacancies
'''
display(pd.read_sql_query(query_4_3, connection))

C:\Users\Leon\AppData\Local\Temp\ipykernel_3528\4206574017.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  display(pd.read_sql_query(query_4_3, connection))


average_minimum  average_maximum
0          71065.0         110537.0

__Вывод: средние значения минимальной и максимальной заработной платы составляют соответственно 71065 и 110537 рублей. Такие суммы зарплат значительно выше средней по стране, и на это может влиять то, что много вакансий расположены в крупных городах. А там зарплаты значительно выше.__

***

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [987]:
# Количество вакансий для каждого сочетания типа рабочего графика.
query_4_4 = f'''select
    schedule,
    employment,
    count(id)
from public.vacancies
group by schedule, employment -- Группировка по типу рабочего графика и типу трудоустройства.
order by count(id) desc -- Обратная сортировка по количеству вакансий.
'''
display(pd.read_sql_query(query_4_4, connection))

C:\Users\Leon\AppData\Local\Temp\ipykernel_3528\530841588.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  display(pd.read_sql_query(query_4_4, connection))


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

__Вывод: самый популярный тип рабочего графика - это полный рабочий день с полной занятостью. На втором месте идет удаленная работа с полной занятостью. Остальные графики сильно отстают от двух перечисленных.__

***

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [988]:
# Количество вакансий в зависимости от требуемого опыта работы.
query_4_5 = f'''select
    experience, -- Опыт работы.
    count(id) -- количество вакансий.
from public.vacancies
group by experience -- Группировка по опыту работы.
order by count(id)
'''
display(pd.read_sql_query(query_4_5, connection))

C:\Users\Leon\AppData\Local\Temp\ipykernel_3528\1690268647.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  display(pd.read_sql_query(query_4_5, connection))


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

__Вывод: в результате запроса по количеству вакансий было выявлено, что порядок требуемого опыта работы по числу этих вакансий следующий: Более 6 лет — Нет опыта — От 3 до 6 лет — От 1 года до 3 лет. По числу вакансий лидирует опыт работы от 1 года до 3 лет.__

***

#### __Выводы по детальному анализу вакансий__
__По результатам детального анализа вакансий удалось выяснить, что основное количество вакансий требуют опыта работы от 1 года до 3 лет с графиком полного рабочего дня. Так же есть много вакансий с удаленным графиком работы. Мало вакансий с требованием опыта работы от 6 лет. Можно предположить, что такие вакансии быстро закрываются, так как люди с таким опытом быстрее находят работу. Средние минимальные зарплаты составляют 71000 рублей, а средние максимальные - 110000 рублей. По количеству вакансий первым идет Московский регион.__

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [989]:
# Работодатели по количеству вакансий.
query_5_1 = f'''select
    e.name employers
from public.vacancies v
join public.employers e on v.employer_id = e.id -- Соединение таблиц с вакансиями и работодателями по ключам.
group by e.name
order by count(v.id) desc
limit 5
'''
display(pd.read_sql_query(query_5_1, connection))

C:\Users\Leon\AppData\Local\Temp\ipykernel_3528\3894954775.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  display(pd.read_sql_query(query_5_1, connection))


employers
0         Яндекс
1     Ростелеком
2       Тинькофф
3           СБЕР
4  Газпром нефть

__Вывод: по запросу можно видеть, что на первом и пятом месте по количеству вакансий стоят компании Яндекс и Газпром нефть соответственно. Так же в пятерку лидеров входят Ростелеком, Тинькофф и СБЕР.__

***

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [990]:
# Находим количество работодателей.
query_5_2 = f'''select
    a.name regions,
    count(distinct e.id) cnt_employers, -- Количество работодателей.
    count(v.id) cnt_vacancies  -- Количество вакансий.
from public.areas a
    left join public.employers e on a.id = e.area -- к таблице с городами присоединяем таблицу с работодателями и таблицу с вакансиями.
    left join public.vacancies v on v.area_id = a.id
where v.id is null -- Выбираем отсутствующие вакансии для более быстрого выполнения запроса.
group by a.name
order by cnt_employers desc
limit 10
'''
display(pd.read_sql_query(query_5_2, connection))

C:\Users\Leon\AppData\Local\Temp\ipykernel_3528\3617172805.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  display(pd.read_sql_query(query_5_2, connection))


regions  cnt_employers  cnt_vacancies
0                 Россия            410              0
1              Казахстан            207              0
2     Московская область             75              0
3     Краснодарский край             19              0
4               Беларусь             18              0
5     Ростовская область             18              0
6            Азербайджан             17              0
7  Нижегородская область             16              0
8   Республика Татарстан             16              0
9             Узбекистан             15              0

__Вывод: Регионом, где наибольшее количество работодателей и при этом нет вакансий, является Россия. Вторым регионом идет Казахстан. Это означает, что в этих регионах работодатели уже нашли нужных сотрудников и закрыли свои вакансии. И это говорит о том, что в этих регионах работодатели очень быстро находят нужных им сотрудников.__

***

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [991]:
# Запрашиваем количество регионов в которых публикуются вакансии работодателей.
query_5_3 = f'''select
    e.name, -- Работодатель.
    count(distinct v.area_id) -- Количество регионов охвата.
from public.vacancies v
    join employers e on v.employer_id = e.id
group by e.name
order by 2 desc
limit 10
'''
display(pd.read_sql_query(query_5_3, connection))

C:\Users\Leon\AppData\Local\Temp\ipykernel_3528\2092942746.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  display(pd.read_sql_query(query_5_3, connection))


name  count
0                      Яндекс    181
1                  Ростелеком    152
2                  Спецремонт    116
3      Поляков Денис Иванович     88
4                   ООО ЕФИН      71
5                  Совкомбанк     63
6                         МТС     55
7  ЭФКО, Управляющая компания     49
8                Почта России     48
9                        КРОН     48

__Вывод: самый большой охват по регионам имеет компания Яндекс. Эта компания имеет вакансии в 181 регионе.__

***

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [992]:
# Запрашиваем работодателей у которых не указана сфера деятельности.
query_5_4 = f'''select
    count(*)
from public.employers_industries i
    right join public.employers e on i.employer_id = e.id
where industry_id is null
'''
display(pd.read_sql_query(query_5_4, connection))

C:\Users\Leon\AppData\Local\Temp\ipykernel_3528\1851649163.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  display(pd.read_sql_query(query_5_4, connection))


count
0   8419

__Вывод: количество работодателей без указания сферы деятельности составляет 8419. Это может быть связано с тем, что сфера деятельности вытекает из названия самой вакансии.__

***

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [993]:
# Запрос по количеству сфер деятельности.
query_5_5 = f'''select
    e.name company_name,
    count(ii.name) cnt_field -- Количество сфер деятельности у компаний.
from public.employers e
    join public.employers_industries i on i.employer_id = e.id
    join public.industries ii on i.industry_id = ii.id
group by e.id
having count(ii.name) = 4 -- После группировки компаний по числу сфер деятельности, выбираем только те, у кого их 4.
order by company_name
offset 2 limit 1 -- Выводим только третье место, как того требует задание.
'''
display(pd.read_sql_query(query_5_5, connection))

C:\Users\Leon\AppData\Local\Temp\ipykernel_3528\2809534890.py:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  display(pd.read_sql_query(query_5_5, connection))


company_name  cnt_field
0         2ГИС          4

__Вывод: на третьем месте, у которой указано 4 сферы деятельности, находится компания 2ГИС.__

***

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [994]:
# Запрос по количеству вакансий в сфере "Разработка программного обеспечения".
query_5_6 = f'''select
    count(e.id)
from public.employers_industries i
    join public.employers e on i.employer_id = e.id
    join public.industries ii on i.industry_id = ii.id
where ii.name = 'Разработка программного обеспечения'
'''
display(pd.read_sql_query(query_5_6, connection))

C:\Users\Leon\AppData\Local\Temp\ipykernel_3528\3476204178.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  display(pd.read_sql_query(query_5_6, connection))


count
0   3553

__Вывод: у 3553 работодателей указана сфера деятельности "Разработка программного обеспечения".__

***

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [995]:
# Запрашиваем количество вакансий в городах-милионниках компании Яндекс.
query_5_7 = f'''select
    a.name,
    count(v.id) cnt
from public.vacancies v
    join public.areas a on v.area_id = a.id
    join employers e on v.employer_id = e.id
where (e.name = 'Яндекс') and (a.name in ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Челябинск', 'Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 'Краснодар', 'Воронеж', 'Пермь', 'Волгоград'))
group by a.name

union all

select
    'Total',
    count(*)
from public.vacancies v
    join public.areas a on v.area_id = a.id
    join employers e on v.employer_id = e.id
where (e.name = 'Яндекс') and (a.name in ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Челябинск', 'Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 'Краснодар', 'Воронеж', 'Пермь', 'Волгоград'))
order by cnt
'''
display(pd.read_sql_query(query_5_7, connection))

C:\Users\Leon\AppData\Local\Temp\ipykernel_3528\3569300033.py:22: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  display(pd.read_sql_query(query_5_7, connection))


name  cnt
0              Омск   21
1         Челябинск   22
2        Красноярск   23
3         Волгоград   24
4             Пермь   25
5            Казань   25
6    Ростов-на-Дону   25
7               Уфа   26
8            Самара   26
9         Краснодар   30
10          Воронеж   32
11      Новосибирск   35
12  Нижний Новгород   36
13     Екатеринбург   39
14  Санкт-Петербург   42
15           Москва   54
16            Total  485

__Вывод: общее количество открытых вакансий у компании Яндекс в городах-миллионниках составляет 485.__

***

### __Выводы по анализу работодателей__

__В первую пятерку лидеров по количеству вакансий входят Яндекс, Ростелеком, Тинькофф, СБЕР и Газпром нефть. Регионами, где много работодателей и много закрытых вакансий, являются Россия и Казахстан. Можно предположить, что в этих регионах нет проблем с поиском сотрудников. У компании Яндекс, которая идет первой в списке по количеству вакансий, поиск ведется в 181 регионе. Следом за Яндекс по числу охвата регионов идут компании Ростелеком и Спецремонт. Так же удалось выснить, что немало вакансий из сферы разработки программного обеспечения.__

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [996]:
# текст запроса
query_6_1 = f'''select
    count(*)
from public.vacancies
where lower(name) like '%data%' or lower(name) like '%данн%'
'''
display(pd.read_sql_query(query_6_1, connection))

C:\Users\Leon\AppData\Local\Temp\ipykernel_3528\3034500601.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  display(pd.read_sql_query(query_6_1, connection))


count
0   1771

__Вывод: количество вакансий, которые имеют отношение к данным, равно 1771.__

***

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [997]:
# Запрос подходящих вакансий для начинающего дата-сайентиста.
query_6_2 = f'''select
    count(*) junior_ds
from public.vacancies
where
    (experience = 'Нет опыта' or employment = 'Стажировка' or name ilike '%junior%') and (
    (name like '%ML%' and name not like '%HTML%') or
    name ilike '%data scientist%' or
    name ilike '%data science%' or
    name ilike '%исследователь данных%' or
    name ilike '%machine learning%' or
    name ilike '%машинн%обучен%'
)'''
display(pd.read_sql_query(query_6_2, connection))

C:\Users\Leon\AppData\Local\Temp\ipykernel_3528\3563588613.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  display(pd.read_sql_query(query_6_2, connection))


junior_ds
0         51

__Вывод: количество подходящих вакансий для начинающего дата-сайентиста равно 51.__

***

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [998]:
# текст запроса
query_6_3 = f'''select
    count(*) cnt_ds
from public.vacancies
where
    (key_skills ilike '%SQL%' or key_skills ilike '%postgres%') and (
    (name like '%ML%' and name not like '%HTML%') or
    name ilike '%data scientist%' or
    name ilike '%data science%' or
    name ilike '%исследователь данных%' or
    name ilike '%machine learning%' or
    name ilike '%машинн%обучен%'
)'''
display(pd.read_sql_query(query_6_3, connection))

C:\Users\Leon\AppData\Local\Temp\ipykernel_3528\3090264066.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  display(pd.read_sql_query(query_6_3, connection))


cnt_ds
0     201

__Вывод: вакансий, где в качестве ключевого навыка указан SQL или postgres, имеется 201 вакансия.__

***

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [999]:
# текст запроса
query_6_4 = f'''select
    count(*) python_ds
from public.vacancies
where
    key_skills ilike '%python%' and (
    (name like '%ML%' and name not like '%HTML%') or
    name ilike '%data scientist%' or
    name ilike '%data science%' or
    name ilike '%исследователь данных%' or
    name ilike '%machine learning%' or
    name ilike '%машинн%обучен%'
)'''
display(pd.read_sql_query(query_6_4, connection))

C:\Users\Leon\AppData\Local\Temp\ipykernel_3528\3219039495.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  display(pd.read_sql_query(query_6_4, connection))


python_ds
0        351

__Вывод: в сфере DS, где требуется знание Python, имеется 351 вакансия.__

***

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [1000]:
# Среднее количество требуемых навыков.
query_6_5 = f'''select
    round(avg((length(key_skills) - length(replace(key_skills, CHR(9), ''))+1)), 2) cnt -- Были подсчитаны табуляций, которые являются разделителями при перечислении навыков и прибавлена единица, т.к. табуляций на единицу меньше при перечислении. 
from public.vacancies
where
    ((name like '%ML%' and name not like '%HTML%') or (
    name ilike '%data scientist%' or
    name ilike '%data science%' or
    name ilike '%исследователь данных%' or
    name ilike '%machine learning%' or
    name ilike '%машинн%обучен%') and
    key_skills is not null
)'''
display(pd.read_sql_query(query_6_5, connection))

C:\Users\Leon\AppData\Local\Temp\ipykernel_3528\1784002172.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  display(pd.read_sql_query(query_6_5, connection))


cnt
0  6.41

__Вывод: среднее значение ключевых навыков равно 6.41.__

***

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [1001]:
# Запрос по средней зарплате для каждого типа требуемого опыта.
query_6_6 = f'''select
    experience,
    round(avg(coalesce((salary_from + salary_to)/2, salary_from, salary_to))) salary
from public.vacancies
where
    ((name like '%ML%' and name not like '%HTML%') or (
    name ilike '%data scientist%' or
    name ilike '%data science%' or
    name ilike '%исследователь данных%' or
    name ilike '%machine learning%' or
    name ilike '%машинн%обучен%')) and
    (salary_from is not null or salary_to is not null)
group by experience
'''
display(pd.read_sql_query(query_6_6, connection))

C:\Users\Leon\AppData\Local\Temp\ipykernel_3528\3875810813.py:16: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  display(pd.read_sql_query(query_6_6, connection))


experience    salary
0           Нет опыта   74643.0
1  От 1 года до 3 лет  139675.0
2       От 3 до 6 лет  243115.0

__Вывод: специалист по DS с опытом работы 3 до 6 лет в среднем может расчитывать на зарплату 243115 рублей.__

***

#### __Выводы по предметному анализу__
__Если брать во внимание вакансии для работы с данными, то таких вакансий 1771. Среднее количество требуемых навыков, которыми нужно обладать специалисту в этой области, равно 6 навыкам. Среди них есть требования по знанию Python (351 вакансия), SQL и postgres (201 вакансия). Для начинающего дата-сайентиста подходит 51 вакансия, где он может рассчитывать на зарплату около 75 тысяч рублей. Если у специалиста DS опыт работы от 1 года до 3 лет, то данный сотрудник может рассчитывать на среднюю зарплату в 140 т.р. Начиная с опыта от 3 лет и выше средняя зарплата может составлять около 243 т.р. и выше.__

# Общий вывод по проекту

__В данном проекте был произведен анализ по вакансиям, работодателям и предметному анализу сфер деятельности этих вакансий. В ходе этого проекта было выявлено, что база насчитывает 294 сферы деятельности, 49197 вакансий и охватывает 1362 региона.
Были выявлены регионы по максимальному количеству открытых вакансий, это города Москва, Санкт-Петербург, Минск, Новосибирск и Алматы.
Произведен глубокий анализ данных в сфере работы с данными. Найдено количество подходящих вакансий для начинающего специалиста по DS. Были рассчитаны для сферы DS средние размеры зарплат в зависимости от стажа и выявлены требования к этим сотрудникам по наличию имеющихся навыков.__